internal/v4/api.go

```diff
@@ -31,25 +42,105 @@ type ReqHandler struct {
 }
 
 func New(pool *charmstore.Pool, config charmstore.ServerParams) Handler {
-	h := Handler{
+	return Handler{
 		Handler: v5.New(pool, config),
 	}
-	// TODO Set h.ResolveURL here.
-	return h
+}
+
+func (h Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// When requests in this handler use router.RelativeURL, we want
+	// the "absolute path" there to be interpreted relative to the
+	// root of this handler, not the absolute root of the web server,
+	// which may be abitrarily many levels up.
+	req.RequestURI = req.URL.Path
+
+	rh, err := h.NewReqHandler()
+	if err != nil {
+		router.WriteError(w, err)
+		return
+	}
+	defer rh.Close()
+	rh.ServeHTTP(w, req)
 }
 
 func NewAPIHandler(pool *charmstore.Pool, config charmstore.ServerParams) charmstore.HTTPCloseHandler {
 	return New(pool, config)
 }
 
+// NewReqHandler fetchs a new instance of ReqHandler
+// from h.Pool and returns it. The ReqHandler must
+// be closed when finished with.
 func (h *Handler) NewReqHandler() (ReqHandler, error) {
-	v5h, err := h.Handler.NewReqHandler()
+	store, err := h.Pool.RequestStore()
 	if err != nil {
-		return ReqHandler{}, errgo.Mask(err, errgo.Is(charmstore.ErrTooManySessions))
+		if errgo.Cause(err) == charmstore.ErrTooManySessions {
+			return ReqHandler{}, errgo.WithCausef(err, params.ErrServiceUnavailable, "")
+		}
+		return ReqHandler{}, errgo.Mask(err)
+	}
+	rh := reqHandlerPool.Get().(ReqHandler)
+	rh.Handler = h.Handler
+	rh.Store = store
+	return rh, nil
+}
+
+func newReqHandler() ReqHandler {
+	h := ReqHandler{
+		ReqHandler: new(v5.ReqHandler),
+	}
+	handlers := v5.RouterHandlers(h.ReqHandler)
+	// TODO mutate handlers appropriately.
+
+	h.Router = router.New(handlers, h)
+	return h
+}
+
+// ResolveURL implements router.Context.ResolveURL,
+// ensuring that any resulting ResolvedURL always
+// has a non-empty PreferredSeries field.
+func (h ReqHandler) ResolveURL(url *charm.URL) (*router.ResolvedURL, error) {
+	return resolveURL(h.Store, url)
+}
+
+// resolveURL implements URL resolving for the ReqHandler.
+// It's defined as a separate function so it can be more
+// easily unit-tested.
```

Comment:
Should there be a comment that this is required by clients depending to have series resolved in the URL?